tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,292 @@
+    return nullptr;
+  }
+  tensorflow::TensorHandle* handle =
+      tensorflow::down_cast<tensorflow::TensorHandleInterface*>(h->handle.get())
+          ->Handle();
+
+  if (handle->IsRemote()) {
+    status->status = tensorflow::errors::InvalidArgument(
+        "DLPack doesn't support remote tensor");
+    return nullptr;
+  }
+  const tensorflow::Tensor* tensor;
+  status->status = handle->Tensor(&tensor);
+  if (!status->status.ok()) {
+    return nullptr;
+  }
+  return tensor;
+};
+
+void DLManagedTensorDeleter(DLManagedTensor* arg) {
+  TFDLManagedTensorCtx* owner =
+      static_cast<TFDLManagedTensorCtx*>(arg->manager_ctx);
+  owner->handle->Unref();
+  delete owner;
+}
+
+DLDataType GetDLDataType(TF_DataType data_type, TF_Status* status) {
+  DLDataType dtype;
+  dtype.lanes = 1;
+  dtype.bits = TF_DataTypeSize(data_type) * 8;
+  switch (data_type) {
+    case TF_DataType::TF_HALF:
+    case TF_DataType::TF_FLOAT:
+    case TF_DataType::TF_DOUBLE:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case TF_DataType::TF_INT8:
+    case TF_DataType::TF_INT16:
+    case TF_DataType::TF_INT32:
+    case TF_DataType::TF_INT64:
+      dtype.code = DLDataTypeCode::kDLInt;
+      break;
+    case TF_DataType::TF_BOOL:
+    case TF_DataType::TF_UINT8:
+    case TF_DataType::TF_UINT16:
+    case TF_DataType::TF_UINT32:
+    case TF_DataType::TF_UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case TF_DataType::TF_BFLOAT16:
+      dtype.code = DLDataTypeCode::kDLBfloat;
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          DataType_Name(static_cast<DataType>(data_type)),
+          " is not supported by dlpack");
+      break;
+  }
+  return dtype;
+}
+
+DLContext GetDLContext(TFE_TensorHandle* h, TF_Status* status) {
+  DLContext ctx;
+  const char* device_name = h->handle->DeviceName(&status->status);
+  DeviceNameUtils::ParsedName parsed_name;
+  tensorflow::DeviceNameUtils::ParseFullName(device_name, &parsed_name);
+  std::string device_type = parsed_name.type;
+  int device_id = -1;
+  if (parsed_name.has_id) {
+    device_id = parsed_name.id;
+  }  // Question: Is it possible that it doens't have id?
+
+  ctx.device_id = device_id;
+  if (device_type == "CPU") {
+    ctx.device_type = DLDeviceType::kDLCPU;
+  } else if (device_type == "GPU") {
+    ctx.device_type = DLDeviceType::kDLGPU;
+  } else {
+    status->status = tensorflow::errors::InvalidArgument(
+        "Unsupported Device Type for dlpack");
+  }
+
+  return ctx;
+}
+
+DLManagedTensor* TFEHandleToTFDLManagedTensorCtx(TFE_TensorHandle* h,
+                                                 TF_Status* status) {
+  const Tensor* tensor = GetTensorFromHandle(h, status);
+  TF_DataType data_type = static_cast<TF_DataType>(tensor->dtype());
+  auto* tf_dlm_tensor_ctx = new TFDLManagedTensorCtx;
+
+  TensorReference* tensor_ref =
+      new TensorReference(*tensor);  // This will call buf_->Ref()
```

Comment:
You could probably save an allocation and just keep a `TensorReference` rather than a `TensorReference*` in the manager context object?

It also appears you leak the `TensorReference` object at the moment?

Reply:
Thanks. I'll change it to `TensorReference`

Reply:
I keep this using `TensorReference*` and manually deleted it at #64
https://github.com/tensorflow/tensorflow/blob/48a353cdaf5697b843aa37299a49201d7f4541e8/tensorflow/c/eager/dlpack.cc#L60-L66

The main reason is that I don't want to add extra constructor on `TFDLManagedTensorCtx` to make it simple

Reply:
> The main reason is that I don't want to add extra constructor on `TFDLManagedTensorCtx`(for `TensorReference`) to make it simple

Honestly, that does not look like a good reason.  Besides the efficiency concern, it is also not clear from reading `TFDLManagedTensorCtx` what the lifetime of the `TensorReference` is.